File: integration/py/read.py
import sys
from typing import NamedTuple, List
import numpy as np

import synnax as sy
from integration import FILE_NAME


# length of channels must = num_iterators
class TestConfig(NamedTuple):
    identifier: str
    num_iterators: int
    chunk_size: int
    bounds: sy.TimeRange
    samples_expected: int
    expected_error: str
    channels: List[List[str]]

    def num_channels(self):
        return sum([len(ch) for ch in self.channels])


client = sy.Synnax(
    host="localhost",
    port=9090,
    username="synnax",
    password="seldon",
    secure=False,
)


class Read_Test:
    _tc: TestConfig

    def __init__(self, argv: List[str]):
        argv_counter = 1
        identifier = argv[argv_counter]
        argv_counter += 1
        num_iterators = int(argv[argv_counter])
        argv_counter += 1
        chunk_size = int(argv[argv_counter])
        argv_counter += 1
        bounds_start = int(argv[argv_counter])
        argv_counter += 1
        bounds_end = int(argv[argv_counter])
        argv_counter += 1
        samples_expected = int(argv[argv_counter])
        argv_counter += 1
        expected_error = argv[argv_counter]
        argv_counter += 1
        number_of_channel_groups = int(argv[argv_counter])
        argv_counter += 1
        channels = []
        for _ in range(number_of_channel_groups):
            number_of_channels_in_group = int(argv[argv_counter])
            argv_counter += 1
            channel_group = []
            for _ in range(number_of_channels_in_group):
                channel_group.append(argv[argv_counter])
                argv_counter += 1
            channels.append(channel_group)

        self._tc = TestConfig(
            identifier=identifier,
            num_iterators=num_iterators,
            chunk_size=chunk_size,
            samples_expected=samples_expected,
            bounds=sy.TimeRange(sy.TimeStamp(bounds_start), sy.TimeStamp(bounds_end)),
            expected_error=expected_error,
            channels=channels,
        )

    def test_with_timing(self):
        start = sy.TimeStamp.now()
        error_assertion_passed = False
        actual_err = "no_error"
        try:
            samples = self.test()
        except Exception as e:
            actual_err = str(e)
            if (
                self._tc.expected_error != "no_error"
                and self._tc.expected_error in str(e)
            ):
                error_assertion_passed = True
            else:
                raise (e)
        else:
            if self._tc.expected_error == "no_error":
                error_assertion_passed = True
        end = sy.TimeStamp.now()

        error_assertion = f"""Expected error: {self._tc.expected_error}; Actual error: {actual_err}\n{"PASS!!" if error_assertion_passed else "FAIL!!"}"""

        s = self.generate_test_report(samples, start.span(end), error_assertion)

        with open(FILE_NAME, "a") as f:
            f.write(s)

    def generate_test_report(
        self, samples: int, time: sy.TimeSpan, error_assertion: str
    ) -> str:
        samples_per_second = samples / (float(time) / float(sy.TimeSpan.SECOND))
        assertion_passed = (
            "PASS!!"
            if (
                self._tc.samples_expected == 0
                or np.isclose(samples, self._tc.samples_expected, rtol=0.01)
            )
            else "FAIL!!"
        )
        assertion_result = (
            f"""Expected samples: {self._tc.samples_expected:,.2f}; Actual samples: {samples:,.2f}\n{assertion_passed}"""
            if self._tc.samples_expected != 0
            else ""
        )

        s = f"""
-- Python Read ({self._tc.identifier})--
Samples read: {samples:,.2f}
Time taken: {time}
Calculated Samples per Second: {samples_per_second:,.2f}
Configuration:
\tNumber of iterators: {self._tc.num_iterators}
\tNumber of channels: {self._tc.num_channels()}
\tChunk size: {self._tc.chunk_size:,.0f}
{assertion_result}
{error_assertion}
"""

        return s

    def test(self) -> int:
        iterators: List[sy.Iterator] = []
        samples_read = 0

        for i in range(self._tc.num_iterators):
            iterators.append(
                client.open_iterator(
                    self._tc.bounds,
                    self._tc.channels[i],
                    self._tc.chunk_size,
                )
            )

        try:
            for i in iterators:
                i.seek_first()
                while i.next(sy.AUTO_SPAN):
                    samples_read += sum([len(s) for s in i.value.series])
        finally:
            for iterator in iterators:
                iterator.close()

        return samples_read


def main():
    Read_Test(sys.argv).test_with_timing()


if __name__ == "__main__":
    main()

File: integration/ts/src/read.ts
import { Synnax } from "@synnaxlabs/client";
import { Iterator } from "@synnaxlabs/client/dist/framer/iterator";
import { CrudeTimeStamp, TimeRange, TimeSpan, TimeStamp } from "@synnaxlabs/x"
import * as fs from 'fs'
import { argv } from 'process';

const FILE_NAME = "../../timing.log"

function approxEqual(a: number, b: number, tol = 0.01): boolean {
    return Math.abs(b - a) < tol * b;
}


class TestConfig {
    identifier: string = "";
    numIterators: number = 0;
    chunkSize: number = 1e5;
    bounds: TimeRange = TimeRange.ZERO;
    samplesExpected: number = 0;
    expectedError: string;
    channels: string[][] = [];

    constructor(
        identifier: string,
        numIterators: number,
        chunkSize: number,
        boundStart: CrudeTimeStamp,
        boundEnd: CrudeTimeStamp,
        samplesExpected: number,
        expectedError: string,
        channels: string[][],
    ) {
        this.identifier = identifier;
        this.numIterators = numIterators;
        this.chunkSize = chunkSize;
        this.bounds = new TimeRange(boundStart, boundEnd);
        this.samplesExpected = samplesExpected;
        this.expectedError = expectedError;
        this.channels = channels;
    }

    numChannels(): number { return this.channels.reduce((a, l) => a + l.length, 0) }
}

const client = new Synnax({
    host: "localhost",
    port: 9090,
    username: "synnax",
    password: "seldon",
    secure: false,
});


class ReadTest {
    tc: TestConfig;
    constructor(argv: string[]) {
        let argvCounter = 2
        const identifier = argv[argvCounter++];
        const numIterators = parseInt(argv[argvCounter++]);
        const chunkSize = parseInt(argv[argvCounter++]);
        const boundStart = BigInt(argv[argvCounter++]);
        const boundEnd = BigInt(argv[argvCounter++]);
        const samplesExpected = parseInt(argv[argvCounter++]);
        const expectedError = argv[argvCounter++];
        const number_of_channel_groups = parseInt(argv[argvCounter++]);
        const channels = [];
        for (let i = 0; i < number_of_channel_groups; i++) {
            const number_of_channels_in_group = parseInt(argv[argvCounter++]);
            const group = [];
            for (let j = 0; j < number_of_channels_in_group; j++) {
                group.push(argv[argvCounter++]);
            }
            channels.push(group);
        }
        this.tc = new TestConfig(identifier, numIterators, chunkSize, boundStart, boundEnd, samplesExpected, expectedError, channels);
    }

    async testWithTiming(): Promise<void> {
        const start = TimeStamp.now();
        let samples = 0;
        let errorAssertion = false;
        let actualError = "";
        let caught = false;
        await this.test().then(result => samples = result).catch((e: unknown) => {
            if (e instanceof Error) {
                caught = true;
                actualError = e.message;
                if (this.tc.expectedError != "no_error" && e.message.includes(this.tc.expectedError)) {
                    errorAssertion = true;
                } else {
                    throw e;
                }
            } else {
                throw e;
            }
        });

        if (!caught) {
            if (this.tc.expectedError == "no_error") {
                errorAssertion = true;
            }
            actualError = "no_error";
        }
        const end = TimeStamp.now();

        const time: TimeSpan = start.span(end);
        const samplesPerSecond = samples / (Number(time) / Number(TimeSpan.SECOND));
        const assertionPassed = this.tc.samplesExpected == 0 || approxEqual(samples, this.tc.samplesExpected);
        const assertionResult = `Expected samples: ${this.tc.samplesExpected}; Actual samples: ${samples}`;
        const s = `
-- TypeScript Read (${this.tc.identifier}) --
Samples read: ${samples}
Time taken: ${time}
Calculated Samples per Second: ${samplesPerSecond.toFixed(2)}
Configuration:
\tNumber of iterators: ${this.tc.numIterators}
\tNumber of channels: ${this.tc.numChannels()}
\tChunk size: ${this.tc.chunkSize}
${assertionResult}: ${assertionPassed ? "PASS!!" : "FAIL!!"}
Expected error: ${this.tc.expectedError}; Actual error: ${actualError}\n${errorAssertion ? "PASS!!" : "FAIL!!"}
`;

        fs.appendFileSync(FILE_NAME, s);
    };

    async test(): Promise<number> {
        const iterators: Iterator[] = new Array(this.tc.numIterators).fill(null);
        let samples_read = 0;

        for (let i = 0; i < this.tc.numIterators; i++) {
            iterators[i] = await client.openIterator(
                this.tc.bounds,
                this.tc.channels[i],
                { chunkSize: this.tc.chunkSize },
            );
        }

        try {
            for (const i of iterators) {
                await i.seekFirst();
                for await (const frame of i) {
                    samples_read += frame.series.reduce((a, s) => a + s.length, 0);
                }
            }
        } finally {
            for (const i of iterators) {
                await i.close();
            }
        }

        return samples_read;
    }
}


async function main() {
    try {
        await new ReadTest(argv).testWithTiming();
    } finally {
        client.close();
    }
}

await main()
